### PR TITLE
Implement importer

### PR DESCRIPTION
**Pull request by enkore**

### CODE REVIEW

File: src/borg_import/main.py [hunk 1]
```diff
@@ -1,17 +1,82 @@
...
+                fd.write('Original path: %s\n' % snapshot_original_path)
+
+            snapshot_original_path.rename(import_path)
+
+            try:
+                borg_import(args, rsnapshot['name'], import_path, timestamp=timestamp)
+            finally:
+                log.debug('  Moving {} -> {}'.format(import_path, rsnapshot['path']))
+                import_path.rename(snapshot_original_path)
+                import_journal.unlink()
```
Comment (enkore) [inline src/borg_import/main.py]: Probably want an option to not do that. Trading "doesn't fiddle with the import source" and "it's fukkken fast!"

File: src/borg_import/main.py [hunk 2]
```diff
@@ -1,17 +1,99 @@
...
 
     if args.rsnapshot_root:
+        import_path = args.rsnapshot_root / 'borg-import-dir'
+        import_journal = args.rsnapshot_root / 'borg-import-dir.snapshot'
+        if import_path.exists():
+            print('{} exists. Cannot continue.'.format(import_path))
+            return 1
         for rsnapshot in get_snapshots(args.rsnapshot_root):
-            print(rsnapshot)
+            timestamp = rsnapshot['timestamp'].replace(microsecond=0)
```
Comment (enkore) [inline src/borg_import/main.py]: Borg should probably be able to accept --timestamps with microsecond resolution.

File: setup.py [hunk 3]
```diff
@@ -46,7 +46,7 @@
     zip_safe=False,
     entry_points={
         'console_scripts': [
-            'borg-import = borg_import.main:main',
+            'borg-import = borg_import.main:below_main',
```
Comment (ThomasWaldmann) [inline setup.py]: wat?

File: src/borg_import/helpers/discover.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 def discover(root, depth):
     """
-    recurse starting from <root> path and yield relative dir pathes with wanted <depth>.
+    recurse starting from <root> path and yield relative dir paths with wanted <depth>.
```
Comment (ThomasWaldmann) [inline src/borg_import/helpers/discover.py]: gotcha :)

File: src/borg_import/helpers/testsuite/test_timestamps.py [hunk 5]
```diff
@@ -1,18 +1,19 @@
 import os
 from datetime import datetime
+from pathlib import Path
```
Comment (ThomasWaldmann) [inline src/borg_import/helpers/testsuite/test_timestamps.py]: seems rather like general refactoring than "implement importer".

File: src/borg_import/main.py [hunk 6]
```diff
@@ -1,17 +1,160 @@
...
 
 from .rsnapshots import get_snapshots
 
+log = logging.getLogger(__name__)
+
+
+def borg_import(args, archive_name, path, timestamp=None):
+    borg_cmdline = ['borg', 'create']
+    if timestamp:
+        borg_cmdline += '--timestamp', timestamp.isoformat()
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]:
```
[1,2]+(3,4)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can only concatenate list (not "tuple") to list
```

File: src/borg_import/main.py [hunk 7]
```diff
@@ -1,17 +1,160 @@
...
+        borg_cmdline.append(args.create_options)
+
+    repository = args.repository.resolve()
+    location = '{}::{}'.format(repository, archive_name)
+    borg_cmdline.append(location)
+    borg_cmdline.append('.')
+
+    log.debug('Borg command line: %r', borg_cmdline)
+    log.debug('Borg working directory: %s', path)
+    subprocess.check_call(borg_cmdline, cwd=str(path))
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]: that will raise even on warnings. which is maybe even ok for a converter.

File: src/borg_import/main.py [hunk 8]
```diff
@@ -1,17 +1,160 @@
...
+class rsnapshotImporter(Importer):
+    name = 'rsnapshot'
+    description = 'import rsnapshot backups'
+
+    def populate_parser(self, parser):
+        parser.add_argument('--backup-set', help='Only consider given backup set (can be given multiple times).',
+                            action='append', dest='backup_sets')
+        parser.add_argument('rsnapshot_root', metavar='RSNAPSHOT_ROOT',
+                            help='Path to rsnapshot root directory', type=Path)
+        # TODO: support the full wealth of borg possibilities
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]: "options" or "features" rather than "possibilities"?

File: src/borg_import/main.py [hunk 9]
```diff
@@ -1,17 +1,160 @@
...
+    log.debug('Borg command line: %r', borg_cmdline)
+    log.debug('Borg working directory: %s', path)
+    subprocess.check_call(borg_cmdline, cwd=str(path))
+
+
+def list_borg_archives(args):
+    borg_cmdline = ['borg', 'list', '--short']
+    repository = args.repository.resolve()
+    borg_cmdline.append(str(repository))
+    return subprocess.check_output(borg_cmdline).decode().split('\n')
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]:
```
>>> s="""a
... b
... c
... """
>>> s.split('\n')
['a', 'b', 'c', '']
>>> s.splitlines()
['a', 'b', 'c']
```

File: src/borg_import/main.py [hunk 10]
```diff
@@ -1,17 +1,160 @@
...
+                print('as', archive_name)
+            else:
+                print()
+            log.debug('  Moving {} -> {}'.format(rsnapshot['path'], import_path))
+
+            with import_journal.open('w') as fd:
+                fd.write('Current snapshot: %s\n' % rsnapshot['name'])
+                fd.write('Original path: %s\n' % snapshot_original_path)
+
+            snapshot_original_path.rename(import_path)
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]: add a comment about why you do this rename?

File: src/borg_import/main.py [hunk 11]
```diff
@@ -1,17 +1,160 @@
...
+                log.debug('  Moving {} -> {}'.format(import_path, rsnapshot['path']))
+                import_path.rename(snapshot_original_path)
+                import_journal.unlink()
+
 
 def main():
-    parser = argparse.ArgumentParser()
-    parser.add_argument("--rsnapshot-root", help="Path to rsnapshot root directory")
+    if not shutil.which('borg'):
+        print('Borg does not seem to be installed. Please install Borg first.')
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]: Maybe rather:

"The 'borg' command can't be found in the PATH. Please correctly install borgbackup first."

File: src/borg_import/main.py [hunk 12]
```diff
@@ -1,17 +1,160 @@
...
+        print('See instructions at https://borgbackup.readthedocs.io/en/stable/installation.html')
+        return 1
+
+    common_parser = argparse.ArgumentParser(add_help=False)
+    common_group = common_parser.add_argument_group('Common options')
+
+    common_group.add_argument("--create-options", "-o",
+                              help="Additional borg create options "
+                                   "(note: Use -o=\"--foo --bar\" syntax to avoid parser confusion).")
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]: good note.

File: src/borg_import/main.py [hunk 13]
```diff
@@ -1,17 +1,160 @@
...
+
+    common_group.add_argument("--debug", action='store_const', dest='log_level', const=logging.DEBUG,
+                              help='Display debug/trace messages.')
+
+    parser = argparse.ArgumentParser(description='Import existing backups from other software to Borg')
+    parser.set_defaults(log_level=logging.WARNING)
+
+    subparsers = parser.add_subparsers()
+
+    for importer in (importer() for importer in Importer.__subclasses__()):
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]: why not simpler like:
```
for importer_class in Importer.__subclasses__():
    importer = importer_class()
    ...
```

File: src/borg_import/main.py [hunk 14]
```diff
@@ -1,17 +1,160 @@
...
+        subparser = subparsers.add_parser(importer.name, help=importer.description, parents=[common_parser])
+        importer.populate_parser(subparser)
+
     args = parser.parse_args()
+    logging.basicConfig(level=args.log_level, format='%(message)s')
 
-    if args.rsnapshot_root:
-        for rsnapshot in get_snapshots(args.rsnapshot_root):
-            print(rsnapshot)
+    if 'function' not in args:
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]: hmm, is args also a dict so no hasattr() is needed?

File: src/borg_import/main.py [hunk 15]
```diff
@@ -1,17 +1,160 @@
...
+        return parser.print_help()
+    return args.function(args)
+
+
+def below_main():
+    # We Don't Go To Ravenholm
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]: too much halflife? :)

File: src/borg_import/main.py (hunk 6 above)
Comment (enkore) [inline src/borg_import/main.py]: Doesn't apply to `+=` / `extend`

Comment (ThomasWaldmann) [inline src/borg_import/main.py]: ah, interesting.

File: setup.py (hunk 3 above)
Comment (enkore) [inline setup.py]: well it's physically correct - below_main is literally below main - but I'm mostly referring to the C runtime meaning here, the functions actually running before main.

but with the meanwhile refactorings it isn't needed anymore

File: src/borg_import/main.py (hunk 14 above)
Comment (enkore) [inline src/borg_import/main.py]: argparse.Namespace is special that way

**Comment by enkore** [comment]
btw. I think if we're really (ever) add another importer then it would probably be better to just make them different commands altogether, `borg-import-rsnapshot`, `borg-import-obnam`, ... and so on.

File: src/borg_import/main.py [hunk 16]
```diff
@@ -25,14 +25,19 @@ def borg_import(args, archive_name, path, timestamp=None):
 
     log.debug('Borg command line: %r', borg_cmdline)
     log.debug('Borg working directory: %s', path)
-    subprocess.check_call(borg_cmdline, cwd=str(path))
+    try:
+        subprocess.check_call(borg_cmdline, cwd=str(path))
+    except subprocess.CalledProcessError as cpe:
+        if cpe.returncode != 1:
+            raise
+        log.debug('Borg exited with a warning (being quiet about it since Borg spoke already)')
```
Comment (ThomasWaldmann) [inline src/borg_import/main.py]: somewhere at the beginning, we need to add "all your data will get assimilated now...". :-)

**Comment by ThomasWaldmann** [comment]
One importer I definitely want to add is for more self-made rsync+hardlinks scripts (like I used personally for a long time). Quite similar to what we have, but not quite.

**Comment by ThomasWaldmann** [comment]
OK, I didn't try it, but assuming it works: collapse some of the commits, merge?

**Comment by enkore** [comment]
It at least works for a basic configuration with two backup sets and about two dozen snapshots across two "generations" (alpha/beta/...). Should be good (for now).

**Comment by ThomasWaldmann** [comment]
maybe we could have a tar with test input data for the different source types an unpack it for the tests.

**Comment by enkore** [comment]
=> #9

**Comment by enkore** [comment]
I don't have write access, you'll need to merge yourself ;)

**Pull request merged**